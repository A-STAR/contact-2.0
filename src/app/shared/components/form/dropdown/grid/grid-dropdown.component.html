<label class="m0">
  <span class="form-label">
    {{ label | translate }}
    <b *ngIf="required" class="text-danger">*</b>
  </span>

  <div appDropdown>
    <div #dropdownTrigger>
      <span [ngClass]="controlClass" [ngStyle]="{ position: 'relative' }" [attr.disabled]="isDisabled || undefined">
        {{ vLabel }}
        <i class="caret"></i>
      </span>
      <input type="hidden" [ngModel]="value">
    </div>
    <div #dropdownContent class="dropdown">
      <ng-container *ngIf="!isDisabled">
        <app-simple-grid
          [columns]="columns"
          [rows]="rows"
          [selection]="selection"
          (select)="onSelect($event)"
        ></app-simple-grid>
        <div class="panel-footer text-right">
          <button class="btn btn-xs btn-default" [disabled]="!value" (click)="onClearClick()">
            {{ 'default.buttons.clear' | translate }}
          </button>
        </div>
      </ng-container>
    </div>
  </div>
</label>
