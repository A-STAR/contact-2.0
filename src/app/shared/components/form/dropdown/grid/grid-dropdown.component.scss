.caret {
  cursor: pointer;
  margin-top: -2px;
  position: absolute;
  right: 10px;
  top: 50%;
}

.dropdown {
  background: #fff;
  border: 1px solid #d5d5d5;
  z-index: 10000;
}
