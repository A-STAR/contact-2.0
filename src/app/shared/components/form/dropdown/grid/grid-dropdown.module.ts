import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { TranslateModule } from '@ngx-translate/core';

import { DropdownModule } from '../../../dropdown/dropdown.module';
import { GridsModule } from '../../../grids/grids.module';

import { GridDropdownComponent } from './grid-dropdown.component';

@NgModule({
  imports: [
    CommonModule,
    DropdownModule,
    FormsModule,
    GridsModule,
    ReactiveFormsModule,
    TranslateModule,
  ],
  exports: [
    GridDropdownComponent,
  ],
  declarations: [
    GridDropdownComponent,
  ],
})
export class GridDropdownModule {}
