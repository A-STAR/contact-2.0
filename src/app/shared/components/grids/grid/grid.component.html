<div class="flex vertical">
  <div *ngIf="toolbar" class="flex-item">
    <app-toolbar-2 [items]="toolbar"></app-toolbar-2>
  </div>

  <div class="flex-item grow">
    <ag-grid-angular
      *ngIf="colDefs"
      class="ag-theme-material"
      style="width: 100%; height: 100%;"
      [treeData]="treeData"
      [animateRows]="true"
      [groupDefaultExpanded]="groupDefaultExpanded"
      [getDataPath]="getDataPath"
      [getRowNodeId]="getRowNodeId"
      [autoGroupColumnDef]="autoGroupColumnDef"
      [columnDefs]="colDefs"
      [getRowClass]="rowClassCallback"
      [gridOptions]="gridOptions"
      [rowData]="rows"
      [rowSelection]="selectionType"
      (gridReady)="onGridReady($event)"
    ></ag-grid-angular>
  </div>

  <app-grid-toolbar
    *ngIf="showToolbar && gridApi"
    class="flex-item"
    [gridApi]="gridApi"
  ></app-grid-toolbar>
</div>

<!--
<ag-grid-angular
  *ngIf="columns"
  class="ag-theme-material"
  style="height: 500px;"
  [columnDefs]="columns"
  [rowData]="rows"
  [deltaRowDataMode]="true"
  [treeData]="true"
  [animateRows]="true"
  [groupDefaultExpanded]="groupDefaultExpanded"
  [getDataPath]="getDataPath"
  [getRowNodeId]="getRowNodeId"
  [autoGroupColumnDef]="autoGroupColumnDef"
  [rowSelection]="rowSelection"
  [gridOptions]="gridOptions"
  (cellValueChanged)="onCellValueChanged($event)"
  (gridReady)="onGridReady($event)"
  (selectionChanged)="onSelect($event)"
  (rowDoubleClicked)="onDblClick($event)"
  (rowDragEnd)="onDragEnd($event)"
></ag-grid-angular>
-->
