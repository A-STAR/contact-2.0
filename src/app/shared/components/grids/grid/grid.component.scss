.toolbar-content {
  align-items: center;
  display: flex;
  height: 100%;
  justify-content: flex-end;
  width: 100%;
}

.quick-filter {
  padding: 0 10px;
  width: 240px;
}

app-simple-grid {
  .toolbar {
    padding: 6px 0 4px !important;
    border-bottom: 1px solid #eaeaea;
  }
}
