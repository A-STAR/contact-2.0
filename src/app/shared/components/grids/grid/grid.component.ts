import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { GridApi } from 'ag-grid';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  encapsulation: ViewEncapsulation.None,
  selector: 'app-simple-grid',
  styleUrls: [ './grid.component.scss' ],
  templateUrl: './grid.component.html'
})
export class SimpleGridComponent<T> {
  @Input() columns;

  @Input('rows') set rows(rows: T[]) {
    this._rows = rows;
    this.updateRows();
  }

  private gridApi: GridApi;

  private _rows: T[];

  get rows(): T[] {
    return this._rows;
  }

  onGridReady(params: any): void {
    this.gridApi = params.api;
    this.gridApi.sizeColumnsToFit();
    this.updateRows();
  }

  private updateRows(): void {
    if (this.gridApi && this.rows) {
      this.gridApi.redrawRows();
    }
  }
}
