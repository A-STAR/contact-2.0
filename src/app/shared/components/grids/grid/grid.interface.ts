import { IGridColumn } from '../grids.interface';

export type ISimpleGridColumn<T> = IGridColumn<T>;
