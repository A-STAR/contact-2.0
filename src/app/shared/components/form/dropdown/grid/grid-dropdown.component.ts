import {
  ChangeDetectionStrategy, ChangeDetectorRef, Component,
  EventEmitter, forwardRef, Input, Output, ViewChild
} from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';

import { IGridColumn } from '../../../grid/grid.interface';

import { DropdownDirective } from '../../../dropdown/dropdown.directive';

@Component({
  selector: 'app-grid-dropdown',
  templateUrl: './grid-dropdown.component.html',
  styleUrls: [ './grid-dropdown.component.scss' ],
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => GridDropdownComponent),
      multi: true
    }
  ],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class GridDropdownComponent<T> implements ControlValueAccessor {
  @Input() columns: Array<IGridColumn>;
  @Input() controlClass = 'form-control';
  @Input() rows: Array<T>;

  @Output() select = new EventEmitter<T>();

  @ViewChild(DropdownDirective) dropdown: DropdownDirective;

  private _selection: T;
  private _isDisabled = false;

  constructor(private cdRef: ChangeDetectorRef) {}

  @Input() labelGetter: (row: T) => string = () => null;
  @Input() valueGetter: (row: T) => string = () => null;

  get label(): string {
    return this._selection ? this.labelGetter(this._selection) : null;
  }

  get value(): string {
    return this._selection ? this.valueGetter(this._selection) : null;
  }

  get isDisabled(): boolean {
    return this._isDisabled;
  }

  get selection(): Array<T> {
    return [ this._selection ];
  }

  writeValue(value: string): void {
    this._selection = (this.rows || []).find(row => this.valueGetter(row) === value);
    this.cdRef.markForCheck();
  }

  registerOnChange(fn: Function): void {
    this.propagateChange = fn;
  }

  registerOnTouched(fn: Function): void {
  }

  setDisabledState(isDisabled: boolean): void {
    this._isDisabled = isDisabled;
  }

  onSelect(row: T): void {
    this.setRow(row);
  }

  onClearClick(): void {
    this.setRow(null);
  }

  private setRow(row: T): void {
    this._selection = row;
    this.propagateChange(this.value);
    this.dropdown.close();
    this.select.emit(row);
  }

  private propagateChange: Function = () => {};
}
