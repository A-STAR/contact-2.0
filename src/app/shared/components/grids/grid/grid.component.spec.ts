import { async as Async, ComponentFixture, TestBed } from '@angular/core/testing';
import { TranslateLoader, TranslateModule } from '@ngx-translate/core';
import { Observable } from 'rxjs/Observable';
import { of } from 'rxjs/observable/of';
import { AgGridModule } from 'ag-grid-angular/main';

import { range } from '@app/core/utils/index';

import { SimpleGridComponent } from './grid.component';
import { setTimeout } from 'timers';

class TranslateLoaderMock {
  getTranslation(language: string): Observable<any> {
    return of([]);
  }
}

describe('SimpleGridComponent', () => {
  let fixture: ComponentFixture<SimpleGridComponent<any>>;

  beforeEach(Async(() => {
    TestBed
      .configureTestingModule({
        declarations: [
          SimpleGridComponent,
        ],
        imports: [
          AgGridModule.withComponents({}),
          TranslateModule.forRoot({
            loader: {
              provide: TranslateLoader,
              useClass: TranslateLoaderMock,
            },
          }),
        ],
      })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(SimpleGridComponent);
  });

  it('should render empty grid', () => {
    fixture.detectChanges();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  // it('should render grid with columns and rows', done => {
  //   fixture.detectChanges();
  //   fixture.componentInstance.columns = range(1, 10)
  //     .map(i => ({
  //       field: `field-${i}`,
  //       headerName: `Column ${i}`,
  //     }));
  //   fixture.componentInstance.rows = range(1, 100)
  //     .map(i => range(1, 10).reduce((acc, j) => ({
  //       ...acc,
  //       [`field-${j}`]: `${j}-${i}`,
  //     }), {}));
  //   fixture.detectChanges();
  //   expect(fixture.nativeElement).toMatchSnapshot();
  // });
});
