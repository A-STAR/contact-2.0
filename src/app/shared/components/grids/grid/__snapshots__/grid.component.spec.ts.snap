// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleGridComponent should render empty grid 1`] = `
<div
  id="root0"
  ng-version="5.2.4"
>
  
  


  

</div>
`;
